system-addon/lib/FaviconFeed.jsm

```diff
@@ -0,0 +1,136 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const {utils: Cu} = Components;
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+Cu.importGlobalProperties(["fetch", "URL"]);
+
+const {actionTypes: at} = Cu.import("resource://activity-stream/common/Actions.jsm", {});
+const {PersistentCache} = Cu.import("resource://activity-stream/lib/PersistentCache.jsm", {});
+const {Prefs} = Cu.import("resource://activity-stream/lib/ActivityStreamPrefs.jsm", {});
+const {getDomain} = Cu.import("resource://activity-stream/lib/TippyTopProvider.jsm", {});
+
+XPCOMUtils.defineLazyModuleGetter(this, "PlacesUtils",
+  "resource://gre/modules/PlacesUtils.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "Services",
+  "resource://gre/modules/Services.jsm");
+
+const TIPPYTOP_UPDATE_TIME = 24 * 60 * 60 * 1000; // 24 hours
+
+this.FaviconFeed = class FaviconFeed {
+  constructor() {
+    this.tippyTopLastUpdated = 0;
+    this.cache = new PersistentCache("tippytop", true);
+    this.prefs = new Prefs();
+    this._sitesByDomain = null;
+  }
+
+  async loadCachedData() {
+    const data = await this.cache.get("sites");
+    if (data && data._timestamp) {
+      this._sitesByDomain = data;
+      this.tippyTopLastUpdated = data._timestamp;
+      this.etag = data._etag;
```

Comment:
This separate `this.etag` is probably not necessary anymore as it'll just be a sub-property off of `_sitesByDomain`